gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-account/src/main/java/io/gravitee/am/gateway/handler/account/resources/account/AccountEndpointHandler.java

```diff
@@ -122,6 +150,158 @@ public void updateProfile(RoutingContext routingContext) {
         }
     }
 
+    public void enrollFactor(RoutingContext routingContext) {
+        User user = getUserFromContext(routingContext);
+        factorService.findById(getFactorIdFromBody(routingContext.getBodyAsJson())).subscribe(factor -> {
+            FactorProvider factorProvider = factorManager.get(getFactorIdFromBody(routingContext.getBodyAsJson()));
+            if (factorProvider.needChallengeSending()) {
+                factorProvider.sendChallenge(new FactorContext(applicationContext, routingContext, new Maps.MapBuilder(new HashMap()).build()))
```

Comment:
For the SMS & Email, where do you read the PhoneNumber and the Email ?
In addition you will probably have to call the `checkSecurityFactor(EnrolledFactor securityFactor)`method on [FactorProvider](https://github.com/gravitee-io/gravitee-access-management/blob/master/gravitee-am-factor/gravitee-am-factor-api/src/main/java/io/gravitee/am/factor/api/FactorProvider.java#L33) in order to check the incoming Phone number & Email Address regarding the factor configuration before calling the send method.